src/ucp/api/ucp.h

```diff
@@ -1878,6 +1981,78 @@ ucs_status_ptr_t ucp_tag_send_sync_nb(ucp_ep_h ep, const void *buffer, size_t co
+ *                          in-parameter and number of actual elements received.
+ *                          I.e. output value can be less or equal input value.
+ * @param [in]     datatype Datatype descriptor for the elements in the buffer.
+ * @param [in]     cb       Callback function that is invoked whenever the
+ *                          receive operation is completed and the data is ready
+ *                          in the receive @a buffer. It is important to note
+ *                          that the call-back is only invoked in a case when
+ *                          the operation cannot be completed in place.
+ * @param [in]     flags    Reserved for future use.
+ *
+ * @return UCS_OK               - The receive operation was completed
+ *                                immediately.
+ * @return UCS_PTR_IS_ERR(_ptr) - The receive operation failed.
+ * @return otherwise            - Operation was scheduled for receive. The
+ *                                request handle is returned to the application
+ *                                in order to track progress of the operation.
+ *                                The application is responsible to released the
+ *                                handle using @ref ucp_request_free
+ *                                "ucp_request_free()" routine.
+ */
+ucs_status_ptr_t ucp_stream_recv_nb(ucp_ep_h ep, void *buffer, size_t *count,
+                                    ucp_datatype_t datatype,
+                                    ucp_stream_recv_callback_t cb,
+                                    unsigned flags);
+
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Non-blocking stream receive data.
+ *
+ * This routine receives a raw data on endpoint @ref ep if available that is
+ * described by UCS_STATUS_PTR(_ptr) in return value as a pointer to the data
+ * and @a length. The routine is a non-blocking and therefore returns
+ * immediately.
+ *
+ * @param [in]   ep               UCP endpoint that is used for the receive
+ *                                operation.
+ * @param [out]  length           Length of received data.
+ *
+ * @return UCS_OK               - No received data available on the @a ep.
+ * @return UCS_PTR_IS_ERR(_ptr) - the receive operation failed and
+ *                                UCS_PTR_STATUS(_ptr) indicates an error.
+ * @return otherwise            - The pointer to the data UCS_STATUS_PTR(_ptr)
+ *                                is returned to the application in order to
+ *                                capture internal UCP resources until received
+ *                                data is processed. When data is processed,
+ *                                application is responsible to release the
+ *                                handle using @ref ucp_stream_data_release
+ *                                "ucp_stream_data_release()" routine.
+ */
+ucs_status_ptr_t ucp_stream_recv_data_nb(ucp_ep_h ep, size_t *length);
```

Comment:
I understand the difference now. I don't think it is well documented. We have to document the differences explicitly. Since there is 0 examples, it is not clear how to use the API.